tensorflow_quantum/core/qsim/state_space.cc

```diff
@@ -149,6 +150,58 @@ void StateSpace::SampleState(const int m, std::vector<uint64_t>* samples) {
   }
 }
 
+tensorflow::Status StateSpace::ComputeSampledExpectation(
+    const tfq::proto::PauliSum& p_sum, StateSpace* scratch,
+    float* expectation_value, const int m) {
+  // apply the  gates of the pauliterms to a copy of the wavefunction
+  // and add up expectation value term by term.
+  tensorflow::Status status = tensorflow::Status::OK();
+  for (const tfq::proto::PauliTerm& term : p_sum.terms()) {
+    // catch identity terms
+    if (term.paulis_size() == 0) {
+      *expectation_value += term.coefficient_real();
+      // TODO(zaqqwerty): error somewhere if identities have any imaginary part
+      continue;
+    }
+
+    // Transform state into the measurement basis and sample it
+    Circuit measurement_circuit;
+    status =
+        ZBasisCircuitFromPauliTerm(term, num_qubits_, &measurement_circuit);
+    if (!status.ok()) {
+      return status;
+    }
+    scratch->CopyFrom(*this);
+    status = scratch->Update(measurement_circuit);
+    if (!status.ok()) {
+      return status;
+    }
+    std::vector<uint64_t> state_samples;
+    scratch->SampleState(m, &state_samples);
+
+    // Find qubits on which to measure parity
+    std::vector<unsigned int> parity_bits;
+    for (const tfq::proto::PauliQubitPair& pair : term.paulis()) {
+      unsigned int location;
+      if (!absl::SimpleAtoi(pair.qubit_id(), &location)) {
+        return tensorflow::Status(
+            tensorflow::error::INVALID_ARGUMENT,
+            "Could not parse Pauli term qubit id: " + pair.ShortDebugString());
+      }
+      parity_bits.push_back(location);
+    }
+
+    // Compute the expectation value over the samples
+    int parity_total(0);
+    for (const uint64_t& sample : state_samples) {
+      parity_total += ComputeParity(parity_bits, sample);
```

Comment:
This looks like it is doing unescessary work. ComputeParity will always recompute the mask from `parity_bits` despite it being the same over and over again. Maybe we should merge the ComputeParity function into here to avoid that ? 

Runtime here is `n_terms * (m*time to compute mask + m*bitwise parity cost)`
Could be reduced to `n_terms * (m * bitwise parity cost + time to compute mask)`

Reply:
Added a new inline function to compute the bitmask only once